### DISCUSSION TITLE
Welcome to BNG Blaster Discussions!

### DISCUSSION THREAD

**Discussion by GIC-de** [rtbrick/bngblaster#4] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Jonix71** [comment]
Is there any support for custom variables to use for DHCP Op82 creation?

    "access-line": {
        "agent-remote-id": "DEU.RTBRICK.{session-global}",
        "agent-circuit-id": "0.0.0.0/0.0.0.0 eth 0:{session-global}"
    }

From config example {session-global} is used. I would like be able to simulate access switches to test our DHCP implementation at the same time we are running BNG Blaster.

ex: GigabitEthernet0/0/{port-nr} switchname{switch-nr} 

Under each port-nr+switch-nr  N-clients should be activated of types (Internet,VoIP,IPTV) 
Possibility to configure port-range used 1-24, 1-48 and swith-nr 1-1000.

**Reply by GIC-de** [reply]
Currently only the variables `{session}` and `{session-global}` are supported, but {session} starts from 1 for every access interface / VLAN range. So you could configure the ARI/ACI under the access interface and create one access interface per ethernet interface 1000 VLAN's per access interface. 

https://rtbrick.github.io/bngblaster/interfaces.html

The configuration attributes for username, agent-remote-id and agent-circuit-id support also some variable substitution. The variable {session-global} will be replaced with a number starting from 1 and incremented for every new session. where as the variable {session} is incremented per interface section.

**Reply by GIC-de** [reply]
```json
{
    "access": [
        {
            "interface": "eth1",
            "agent-remote-id": "DEU.RTBRICK.{session-global}",
            "agent-circuit-id": "GigabitEthernet0/0/1 switchname{session}"
            "outer-vlan-min": 100,
            "outer-vlan-max": 199,
            "inner-vlan-min": 7,
            "inner-vlan-max": 7
        },
        {
            "interface": "eth1",
            "agent-remote-id": "DEU.RTBRICK.{session-global}",
            "agent-circuit-id": "GigabitEthernet0/0/2 switchname {session}"
            "outer-vlan-min": 200,
            "outer-vlan-max": 299,
            "inner-vlan-min": 7,
            "inner-vlan-max": 7
        }
    ]
}
```

**Reply by Jonix71** [reply]
Ok,
It works but will generate a large config file. Another question for DHCP part have you thought about adding possibility to send Option-60? Our BNG implementation uses Option-60 to classify clients and this also vill give client diffrent IP allocated from the DHCP server. So that would be a feature request.

**Reply by GIC-de** [reply]
The DHCPv4 option 60 is currently not supported but could be added. I would assume you need an option 60 which is different for each user with some iterators? Could you please open an issue/feature request. There you could also specify an example configuration that would fit your needs, this helps to better understand your request.

**Comment by onfirealot** [comment]
Is there any plans to add LAC into the testing at all? We're currently using Spirent testing, but have started to use BNG Blaster for PPPoE testing and finding it really useful.

**Reply by GIC-de** [reply]
I'm not sure if I understand the question correctly, but the BNG Blaster is already able to emulate [L2TP LNS servers](https://rtbrick.github.io/bngblaster/access/l2tp.html). The assumption is that the LAC is the device under test (DUT). We test our BNG LAC functionality in RtBrick FullStack software (RBFS) using BNG Blaster as well. 

Here the BNG Blaster supports also RFC5515 CSUN and CSURQ.

**Reply by onfirealot** [reply]
Ah ok. We are currently testing the LTS functionality our BNG, so we are using a Spirent tester to act as a LAC. That way we can tunnel switch the incoming subscribers on the BNG onto another device.

**Comment by marthin-vandyk** [comment]
Hi guys,

Am looking to emulate a lot of pppoe sessions, but they must be in the same VLAN. Is there a way to do so without having the VLAN range?

thanks

**Reply by GIC-de** [reply]
You need to set VLAN mode to N:1, in this case the Blaster will not iterate VLAN's and creates multiple sessions per VLAN. 

```
        {
            "interface": "eth1",
            "vlan-mode": "N:1"
            "inner-vlan": 128,
            "outer-vlan": 7,
        }
```

This works single tagged, double tagged, tripple tagged and even untagged.

**Reply by marthin-vandyk** [reply]
Many thanks Christian! I did find the setting in the docs, but was not sure how it applied to the topology and thus did not try it/overlooked it.

did try many other things :)

looking good thanks again

**Reply by GIC-de** [reply]
I created a documentation issue to improve documentation here:
https://github.com/rtbrick/bngblaster/issues/127

**Comment by girishpatoliya** [comment]
Does LNS mode support ppp over L2TP ?
I have tried to run an instance of bng blaster, but I don't see it listening on port 1701..

**Reply by GIC-de** [reply]
Yes, LNS with PPP over L2TPv2 is supported. The BNG Blaster is completely build in user space, therefore you will not see any listening sockets at all. It is also important to not have any IP address configured on the interfaces used by BNG Blaster, otherwise Blaster and Linux will consume the same packet which can lead into some issues. Details see interface section of the documentation.

**Reply by girishpatoliya** [reply]
Thanks for prompt response, I am running bng blaster in ubuntu container with network host,
I don't have option to allocate dedicate interface. I Need to use existing eth2 which is having IPs.
Please share any workaround to use existing interfaces, Or I will have to look for some other simulator.

**Reply by GIC-de** [reply]
It's not recommended but it should work as long as IP addresses are not equal.

**Reply by girishpatoliya** [reply]
after changing config to have dedicated network interface and dummy interface on access, l2tp started working, however bng blaster is not responding to lcp configuration request.
in logs I only see 
```
Apr 27 13:25:00.450083 All network interfaces resolved
Apr 27 13:25:28.176969 Received signal Interrupt (2), initiating teardown
Apr 27 13:25:29.452939 ALL SESSIONS TERMINATED
```
![image](https://user-images.githubusercontent.com/5243247/234873043-58d9a010-0bef-40ea-8370-715eee5b32ef.png)

How can I debug as why PPP is not responded.

**Reply by GIC-de** [reply]
Both PPPoE client and LNS are BNG Blaster, or just LNS? 

Best is to start BNG Blaster with `-P test.pcap` to see what traffic is send by Blaster and received. The Blaster PCAP includes also meta informations like interface and direction. You can also start Blaster with logging `-l debug -l l2tp -l pppoe`.

**Reply by girishpatoliya** [reply]
Just LNS is BNG blaster, LAC is DUT, attached snapshot is from -P test.pcap only.
I will capture debug logs and update.

```
 L2TP Error (LNS1) Wrong challenge response in SCCCN from 192.168.13.100
 ```
 I have used below strings at both sides:
```
     "l2tp-server": [
        {
            "name": "LNS1",
            "address": "192.168.13.15",
            "secret": "passwordc",
            "receive-window-size": 8
        }
```

**Reply by girishpatoliya** [reply]
Challenge issue is resolved after changing secret to alpha-numeric.
However PPP/LCP config request is still not responded (ppp over l2tp)

![image](https://user-images.githubusercontent.com/5243247/234889982-7545a83b-d18d-4dc0-a3dd-28fa193092b2.png)

**Reply by GIC-de** [reply]
I'm confused, your client should be in PPP authentication phase, waiting for PAP or CHAP authentication ACK. So would expect authentication request retries here rather than LCP configuration requests.

**Reply by girishpatoliya** [reply]
I have already tested ppp/lcp/chap/ipcp with openl2tp with same LAC, I wanted to test ipv6cp with BNG.
Sequece is : L2tp tunnel -> PPP LCP -> PAP|CHAP ->IPCP -> data packets. (attached pcap snap from test with openl2tp)

![image](https://user-images.githubusercontent.com/5243247/235066272-928b0733-2771-4c6c-8e21-6a8698b8be8f.png)

**Reply by girishpatoliya** [reply]
I added debug statement (last else) in bbl_l2tp.c: bbl_l2tp_data_rx()
```
 907     switch(l2tp->protocol) {
 908         case PROTOCOL_LCP:
 909             lcp_rx = (bbl_lcp_s*)l2tp->next;
 910             lcp_rx->padding = l2tp_session->tunnel->server->lcp_padding;
 911             if(lcp_rx->code == PPP_CODE_ECHO_REQUEST) {
 912                 lcp_rx->code = PPP_CODE_ECHO_REPLY;
 913                 bbl_l2tp_send_data(l2tp_session, PROTOCOL_LCP, lcp_rx);
 914             } else if(lcp_rx->code == PPP_CODE_TERM_REQUEST) {
 915                 l2tp_session->disconnect_code = 3;
 916                 l2tp_session->disconnect_protocol = 0;
 917                 l2tp_session->disconnect_direction = 1;
 918                 bbl_l2tp_send(l2tp_session->tunnel, l2tp_session, L2TP_MESSAGE_CDN);
 919                 bbl_l2tp_session_delete(l2tp_session);
 920             } else {
 921                     bbl_l2tp_tunnel_s *l2tp_tunnel = l2tp_session->tunnel;
 922                     LOG(L2TP, "**L2TP Info (%s) Tunnel (%u) from %s (%s) session (%u) not handling LCP CONF REQ**\n",
 923                                     l2tp_tunnel->server->host_name, l2tp_tunnel->tunnel_id,
 924                                     l2tp_tunnel->peer_name,
 925                                     format_ipv4_address(&l2tp_tunnel->peer_ip),
 926                                     l2tp_session->key.session_id);
 927             }
 
 Observed in logs when LCP received from LAC.
 ===============================
 
 Apr 28 07:17:40.048377 L2TP Debug (LNS1) Tunnel 1 state changed from Wait-Control-Connection to Established
Apr 28 07:17:40.049093 L2TP Info (LNS1) Tunnel 1 with testlac1 (192.168.13.100) established
Apr 28 07:17:40.061813 L2TP Info (LNS1) Tunnel (1) from testlac1 (192.168.13.100) session (1) established
Apr 28 07:17:40.629333 L2TP Info (LNS1) Tunnel (1) from testlac1 (192.168.13.100) session (1) **not handling LCP CONF REQ**
Apr 28 07:17:43.423148 L2TP Debug (LNS1) Tunnel 1 Session 1 deleted
Apr 28 07:17:43.424064 L2TP Debug (LNS1) Tunnel 1 state changed from Established to Received-StopCCN
Apr 28 07:17:49.033989 L2TP Debug (LNS1) Tunnel 1 state changed from Received-StopCCN to Terminated
 
```

Below is the topology I am using :
                                                               
 (LAC- 192.168.13.100) ------------------> (192.168.13.10 - BNG Blaster)


Config :
```
{
    "interfaces": {
        "network": [
            {
                "interface": "eth4",
                "address": "192.168.13.11/24",
                "gateway": "192.168.13.100"
            }
        ],
        "access": [
            {
                "interface": "dummy0",
                "outer-vlan-min": 1,
                "outer-vlan-max": 4000,
                "inner-vlan-min": 8,
                "inner-vlan-max": 8,
                "authentication-protocol": "CHAP"
            }
        ]
    },
    "ppp": {
        "mru": 1470,
        "authentication": {
            "username": "testlac1",
            "password": "testcal1",
            "timeout": 5,
            "retry": 60
        },
        "lcp": {
            "conf-request-timeout": 5,
            "conf-request-retry": 30,
            "keepalive-interval": 30,
            "keepalive-retry": 3
        },
        "ipcp": {
            "enable": true
        },
        "ip6cp": {
            "enable": true
        }
    },
    "l2tp-server": [
        {
            "name": "LNS1",
            "address": "192.168.13.10",
            "secret": "passwordc",
            "receive-window-size": 8
        }
    ],
    "session-traffic": {
        "autostart": true,
        "ipv4-pps": 1
    }
}

```

**Reply by GIC-de** [reply]
This is not needed for tests with BNG Blaster as client and server but we can extend. Currently I'm busy with other Blaster requests, so if you like, you can try yourself and open PR. Otherwise I will work on it as soon as possible.

**Comment by paramountk** [comment]
Hi, awesome tool - currently building up a config in lab env in prep for testing near go-live solution.
I have a simple question, that I cannot seem to find the answer for:

Is it possible to send traffic streams between the IPOE sessions established and endpoints matching BGP routes injected?
I currently have BGP working, and traffic working, but traffic is to network interface.  I could not seem to work out how to link the two.
Many thanks for any info.

**Reply by GIC-de** [reply]
Yes, this works. You just need to configure the `network-ipv4-address` option for each stream, assigning it one of the BGP-advertised addresses.

**Reply by paramountk** [reply]
Just tried it and all is working, many thanks for swift response.

**Comment by jose-bogo** [comment]
We are using BNGBlaster to test a CGNAT BPA solution, with dynamic block allocation.
In our tests, we were able to occupy only 01 TCP port and 01 CPU port, without increasing.
The ideal would be to occupy all available ports.
Our CGNAT makes 16 blocks of 256 ports available to each subscriber.
Is it possible to do this with BNGBlaster?

**Reply by jose-bogo** [reply]
Our config:
 
{
    "interfaces": {
        "tx-threads": 6,
        "tx-interval": 0.01,
        "rx-threads": 6,
        "rx-interval": 0.1,
        "io-slots": 32768,
        "links": [
            {
                "interface": "enp3s0f1",
                "io-mode": "raw",
                "rx-threads": 6,
                "rx-cpuset": [2],
                "tx-threads": 6,
                "tx-cpuset": [3]
            },
            {
                "interface": "enp3s0f0",
                "io-mode": "raw",
                "rx-threads": 6,
                "rx-cpuset": [4],
                "tx-threads": 6,
                "tx-cpuset": [5]
            }
        ],
        "network": [
            {
                "interface": "enp3s0f1",
                "address": "10.10.1.2/16",
                "gateway": "10.10.1.1"
            }
        ],
        "access": [
            {
                "interface": "enp3s0f0",
                "type": "ipoe",
                "ipv6": false,
		"vlan-mode": "N:1",
		"qinq": false,
		"outer-vlan": 0,
                "inner-vlan": 0,
	        "stream-group-id": 1
	    }
	]
    },
    "dhcp": {
                 "enable": true,
                 "broadcast": true,
		 "retry": 100
    },
    "http-client": [
        {
    	        "http-client-group-id": 1,
        	"name": "CLIENT-1",
                "url": "blaster.rtbrick.com",
                "destination-ipv4-address": "10.10.1.2",
                "destination-port": 80
        }
    ],
    "http-server": [
        {
            "name": "SERVER",
            "ipv4-address": "10.10.0.2",
            "port": 80,
            "network-interface": "enp3s0f1"
        }
    ],
    "sessions": {
        "count": 500,
        "max-outstanding": 1200,
	"start-rate": 1200,
        "stop-rate": 1200,
        "start-delay": 0
    },
    "streams": [
        {
            "stream-group-id": 1,
            "name": "S1",
            "type": "ipv4",
            "direction": "both",
            "priority": 0,
            "nat": true,
            "length": 1280,
            "pps": 25000,
            "network-interface": "enp3s0f1",
            "start-delay": 5,
	    "source-port": 1025,
	    "destination-port": 65535
        },
	{
            "name": "TCP1",
            "stream-group-id": 1,
            "type": "ipv4",
            "direction": "both",
            "pps": 100,
            "setup-interval": 30,
	    "nat": true,
            "raw-tcp": true,
            "network-ipv4-address": "10.10.0.2"
        }
    ]
}

**Reply by jose-bogo** [reply]
![image](https://github.com/user-attachments/assets/ca304417-265a-46b9-b594-f7bbd8094664)

**Reply by GIC-de** [reply]
Yes, that's possible. We use it internally to test 4 million NAT entries in our Broadcom (Q2c)-based NAT solution. To achieve this, you simply need to define more streams per user, 256 streams in this case, to fully utilize the assigned block.

When setting up millions of streams, you’ll need to reduce the PPS per stream to keep the overall PPS within acceptable limits. For instance, you can set the PPS to less than 1, such as 0.1 PPS, which sends one packet every 10 seconds, allowing for 10 million flows with 1 million PPS.

For more details, visit: https://rtbrick.github.io/bngblaster/nat.html#](https://rtbrick.github.io/bngblaster/nat.html#

**Reply by jose-bogo** [reply]
I changed the configs, but the only that I noticed was the decrease of bandwidth usage.
Any tip?
This is my config updated:
.
.
.
    "streams": [
        {
            "stream-group-id": 1,
            "name": "S1",
            "type": "ipv4",
            "direction": "both",
            "priority": 0,
            "nat": true,
            "length": 1280,
            "pps": 1,
	    "setup-interval": 1,
            "network-interface": "enp3s0f1",
            "start-delay": 5,
	    "source-port": 1025,
	    "destination-port": 65535
        },
	{
            "name": "TCP1",
            "stream-group-id": 1,
            "type": "ipv4",
            "direction": "both",
            "pps": 1,
            "setup-interval": 1,
	    "nat": true,
            "raw-tcp": true,
            "network-ipv4-address": "10.10.0.2"
        }
    ],
    "traffic":[
         { 
 	"stream-rate-calculation": false 
       	 },
	 { 
	 "stream-delay-calculation": false
	 }
    ]
}

**Comment by d4nielsan** [comment]
Hi, 

I am looking generate DHCPv4 traffic (IPoE), in order to test a DHCP server. The server and the BNGBlaster are on the the same LAN (so no relays or other intermiate devices). I ran the command `sudo bngblaster -C dhcp.json -I --logging dhcp`, and can see DHCP-discover messages, but nothing further. The configured session never establishes. Can you point me to where I am going wrong? Thanks for your advice. My network connections are shown below:

```
# /etc/netplan/50-cloud-init.yaml 
network:
    ethernets:
        enp1s0:
            dhcp4: no
            addresses:
            - 192.168.122.22/24
            routes:
            - to: default
              via: 192.168.122.1
            nameservers:
              addresses: [8.8.8.8, 8.8.4.4]
        ens1:
          dhcp4: no
          dhcp6: no
          link-local: []
    version: 2

```

...And my JSON file
```
# dhcp.jsom
{
    "sessions": {
        "count": 1,
        "reconnect": true,
        "autostart": true,
        "start-delay": 2,
        "start-rate": 200,
        "stop-rate": 200,
        "max-outstanding": 20000
    },
    "interfaces": {
        "access": [
            {
                "__comment__": "DHCP Server",
                "interface": "ens1",
                "type": "ipoe",
                "vlan-mode": "1:1",
                "outer-vlan-min": 1001,
                "outer-vlan-max": 1002,
                "inner-vlan-min": 1000,
                "inner-vlan-max": 3000
            }
        ]
    },
    "access-line": {
        "agent-remote-id": "PROVIDER{session-global}",
        "agent-circuit-id": "PROVIDER{session-global}"
    },
    "ipoe": {
        "arp-interval": 20,
        "arp-timeout": 1
    },
    "dhcp": {
        "enable": true,
        "broadcast": true,
        "timeout": 5,
        "retry": 5,
        "access-line": true
    },
    "dhcpv6": {
        "enable": false
    }
}`
```
